File: kernel/tests/recycle_bin.rs
mod shared;
mod workspace;

use almond_kernel::{
    adapters::{
        meta::RequestMeta,
        recycle_bin::{CreateRecycleBinEntry, RecycleBinItemType},
    },
    error::KernelError,
    repositories::prelude::RecycleBinRepositoryExt,
};

use fake::{
    Fake,
    faker::lorem::en::Paragraph,
};

use uuid::Uuid;

use shared::*;

#[tokio::test]
async fn test_store_without_workspace_recycle_bin() -> Result<(), KernelError> {
    let repo = get_recycle_bin_repository().await;

    let payload = CreateRecycleBinEntry {
        item_id: Uuid::new_v4(),
        item_type: RecycleBinItemType::Note,
        payload: Paragraph(1..2).fake(),
        workspace_identifier: None,
    };

    let resp = repo.store(&payload, &None::<RequestMeta>).await;

    assert!(resp.is_err());

    Ok(())
}

#[tokio::test]
async fn test_store_with_workspace_recycle_bin() -> Result<(), KernelError> {
    let (meta, repo) = setup_workspace(get_recycle_bin_repository).await?;

    let payload = CreateRecycleBinEntry {
        item_id: Uuid::new_v4(),
        item_type: RecycleBinItemType::Note,
        payload: Paragraph(1..2).fake(),
        workspace_identifier: Some(meta.workspace_identifier),
    };

    let entry = repo.store(&payload, &Some(meta.clone())).await?;

    assert_eq!(entry.item_id, payload.item_id);
    assert_eq!(entry.workspace_identifier, Some(meta.workspace_identifier));

    Ok(())
}

#[tokio::test]
async fn test_find_by_id_recycle_bin() -> Result<(), KernelError> {
    let (meta, repo) = setup_workspace(get_recycle_bin_repository).await?;

    let payload = CreateRecycleBinEntry {
        item_id: Uuid::new_v4(),
        item_type: RecycleBinItemType::Note,
        payload: Paragraph(1..2).fake(),
        workspace_identifier: Some(meta.workspace_identifier),
    };

    let created = repo.store(&payload, &Some(meta.clone())).await?;

    let found = repo
        .find_by_id(&created.identifier, &Some(meta.clone()))
        .await?;

    assert!(found.is_some());
    assert_eq!(found.unwrap().identifier, created.identifier);

    Ok(())
}

#[tokio::test]
async fn test_find_all_recycle_bin() -> Result<(), KernelError> {
    let (meta, repo) = setup_workspace(get_recycle_bin_repository).await?;

    for _ in 0..3 {
        let payload = CreateRecycleBinEntry {
            item_id: Uuid::new_v4(),
            item_type: RecycleBinItemType::Note,
            payload: Paragraph(1..2).fake(),
            workspace_identifier: Some(meta.workspace_identifier),
        };

        repo.store(&payload, &Some(meta.clone())).await?;
    }

    let results = repo.find_all(&Some(meta.clone())).await?;

    assert!(results.len() >= 3);

    Ok(())
}

#[tokio::test]
async fn test_find_by_item_type_recycle_bin() -> Result<(), KernelError> {
    let (meta, repo) = setup_workspace(get_recycle_bin_repository).await?;

    for _ in 0..3 {
        let payload = CreateRecycleBinEntry {
            item_id: Uuid::new_v4(),
            item_type: RecycleBinItemType::Note,
            payload: Paragraph(1..2).fake(),
            workspace_identifier: Some(meta.workspace_identifier),
        };

        repo.store(&payload, &Some(meta.clone())).await?;
    }

    let results = repo
        .find_by_item_type(&RecycleBinItemType::Note, &Some(meta.clone()))
        .await?;

    assert!(results.len() >= 3);

    Ok(())
}

#[tokio::test]
async fn test_purge_recycle_bin() -> Result<(), KernelError> {
    let (meta, repo) = setup_workspace(get_recycle_bin_repository).await?;

    let payload = CreateRecycleBinEntry {
        item_id: Uuid::new_v4(),
        item_type: RecycleBinItemType::Note,
        payload: Paragraph(1..2).fake(),
        workspace_identifier: Some(meta.workspace_identifier),
    };

    let created = repo.store(&payload, &Some(meta.clone())).await?;

    repo.purge(&created.identifier, &Some(meta.clone()))
        .await?;

    let found = repo
        .find_by_id(&created.identifier, &Some(meta.clone()))
        .await?;

    assert!(found.is_none());

    Ok(())
}

#[tokio::test]
async fn test_purge_all_recycle_bin() -> Result<(), KernelError> {
    let (meta, repo) = setup_workspace(get_recycle_bin_repository).await?;

    for _ in 0..3 {
        let payload = CreateRecycleBinEntry {
            item_id: Uuid::new_v4(),
            item_type: RecycleBinItemType::Note,
            payload: Paragraph(1..2).fake(),
            workspace_identifier: Some(meta.workspace_identifier),
        };

        repo.store(&payload, &Some(meta.clone())).await?;
    }

    repo.purge_all(&Some(meta.clone())).await?;

    let results = repo.find_all(&Some(meta.clone())).await?;

    assert!(results.is_empty());

    Ok(())
}

File: nuxt-app/app/utils/lunar/repositories/recycle_bin.ts
import type { CreateRecycleBinEntry, ItemType, RecycleBin } from "lunar";
import { BaseRepository, type RequestMeta } from "../base";

export type { CreateRecycleBinEntry };

export class RecycleBinRepository extends BaseRepository {
  async store(
    payload: CreateRecycleBinEntry,
    meta?: RequestMeta,
  ): Promise<RecycleBin> {
    const m = this.requireMeta(meta);
    return this.mustRow<RecycleBin>(
      `INSERT INTO recycle_bin
         (identifier, item_id, item_type, payload, deleted_at, workspace_identifier)
       VALUES ($1, $2, $3, $4, $5, $6)
       RETURNING *`,
      [
        this.newUuid(),
        payload.itemId,
        payload.itemType,
        payload.payload,
        this.now(),
        m.workspaceIdentifier,
      ],
    );
  }

  async find_by_id(
    identifier: string,
    meta?: RequestMeta,
  ): Promise<RecycleBin | null> {
    const m = this.requireMeta(meta);
    return this.row<RecycleBin>(
      `SELECT * FROM recycle_bin WHERE identifier = $1 AND workspace_identifier = $2`,
      [identifier, m.workspaceIdentifier],
    );
  }

  async find_all(meta?: RequestMeta): Promise<RecycleBin[]> {
    const m = this.requireMeta(meta);
    return this.rows<RecycleBin>(
      `SELECT * FROM recycle_bin WHERE workspace_identifier = $1 ORDER BY deleted_at DESC`,
      [m.workspaceIdentifier],
    );
  }

  async find_by_item_type(
    item_type: ItemType,
    meta?: RequestMeta,
  ): Promise<RecycleBin[]> {
    const m = this.requireMeta(meta);
    return this.rows<RecycleBin>(
      `SELECT * FROM recycle_bin WHERE item_type = $1 AND workspace_identifier = $2 ORDER BY deleted_at DESC`,
      [item_type, m.workspaceIdentifier],
    );
  }

  async purge(identifier: string, meta?: RequestMeta): Promise<void> {
    const m = this.requireMeta(meta);
    await this.run(
      `DELETE FROM recycle_bin WHERE identifier = $1 AND workspace_identifier = $2`,
      [identifier, m.workspaceIdentifier],
    );
  }

  async purge_all(meta?: RequestMeta): Promise<void> {
    const m = this.requireMeta(meta);
    await this.run(`DELETE FROM recycle_bin WHERE workspace_identifier = $1`, [
      m.workspaceIdentifier,
    ]);
  }
}
